cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/common/plugin/WrappedReduceAggregator.java

```diff
@@ -118,7 +158,16 @@ public AGG_VALUE initializeAggregateValue(GROUP_VALUE val) throws Exception {
   public AGG_VALUE mergeValues(AGG_VALUE aggValue, GROUP_VALUE value) throws Exception {
     operationTimer.start();
     try {
-      return caller.call(() -> aggregator.mergeValues(aggValue, value));
+      return caller.call((Callable<? extends AGG_VALUE>) () -> {
+        try {
+          return aggregator.mergeValues(aggValue, value);
+        } catch (Exception e) {
+          if (caller instanceof StageLoggingCaller) {
```

Comment:
instanceof is often an indication that logic is happening in the wrong place. For example, we may refactor things in the future that makes this caller some delegate class instead of StageLoggingCaller and unintentionally break this.

The desired functionality looks like it can be more cleanly done by creating something like an ExceptionWrappingCaller that takes a delegate Caller, catches any exception and wraps it into a WrappedStageException before rethrowing it. Something like:

```
public ExceptionWrappingCaller<T> extends Caller<T> {
  private final Caller delegate;
  private final String stageName;

  public <T> T call(Callable<T> callable) throws Exception {
    try {
      return delegate.call();
    } catch (Exception e) {
      throw new WrappedStageException(e, stageName);
    }
  }
}
```


Then wherever these Wrapped plugins are created, pass that caller in. This would also remove a lot of duplicate code in this PR.

Reply:
Thanks for the comment, this makes the code look more cleaner.

Acknowledged.